dev-docs/RFCs/v8.0/carto-v9-rfc.md

```diff
@@ -0,0 +1,222 @@
+- In v9 of deck.gl we will have Typescript.
+
+As the CARTO platform has grown, the CartoLayer has been used to implement a wide variety of functions, with the result that it can be unclear to developers what is going on:
+
+- The primitive layer used for rendering is obscured from the user, e.g if `geoColumn=geom` it will use a `GeoJsonLayer`, if `geoColumn=h3|quadbin` `H3HexagonLayer` or `QuadbinLayer` will be used.
+- Props are dependent on the layer type being used, e.g. `aggregationExp` has no meaning for non-spatial index datasets. `QueryParameters` are only available for `Query` layers.
+- No easy way for the developer to use a custom layer for rendering.
+- No types to enforce correct options.
+
+We are working right now in v9, so I think it's a great time to address those changes. Especially because v9 will expose types at the package roots.
+
+## Background
+
+The current implementation of `CartoLayer` is as a `CompositeLayer` which encapsulates all the data loading via the CARTO API, and depending on the particular configuration of props renders the data using one of the following layers:
+
+- `CartoTileLayer`
+- `H3TileLayer`
+- `MVTLayer`
+- `RasterTileLayer`
+- `QuadbinTileLayer`
+
+The standard pattern in deck.gl is for a `data` prop of `Layer` to accept the actual data to be visualized, or a Promise that resolves to the data. The `MVTLayer` additionally [accepts TileJSON](https://deck.gl/docs/api-reference/geo-layers/mvt-layer#data) as input, as do all the CARTO Layers.
+
+## API background
+
+Without getting into the details too much, using the current `CartoLayer` with the CARTO works in the following way:
+
+1. User configures `CartoLayer` with a set of props
+2. `CartoLayer` uses these props to construct a URL for a Map Instantiation request
+3. `CartoLayer` performs a Map Instantiation fetch request and obtains URLs for data endpoints, in different formats: TileJSON, GeoJSON, JSON etc
+4. `CartoLayer` is [hardcoded](https://github.com/visgl/deck.gl/blob/master/modules/carto/src/layers/carto-layer.ts#L263) to only use TileJSON, so the TileJSON URL is fetched
+5. `CartoLayer` [decides](https://github.com/visgl/deck.gl/blob/master/modules/carto/src/api/layer-map.ts#L203-L223) which Layer to use for rendering and passes the TileJSON payload to that layer in the `data` prop
+
+## Proposal
+
+At a high level we want to achieve an API which makes it clearer to the user what is going on, and allows for better extensibility. We also want to separate the options used for loading the data from the props which style the layer.
+
+To do so, the interaction with the CARTO API will be done via a set of typed helper classes which aid the user in obtaining the TileJSON needed for a particular layer.
+
+```javascript
+import CartoQuerySource, CartoQuerySourceOptions, CartoTileLayer from '@deck.gl/carto';
+
+// Global options
+setDefaultCredentials({accessToken: 'XXXX'});
```

Comment:
I think we should deprecated this too. The credentials are storing two things today: accessToken and apiBaseUrl. The apiBaseUrl is not a `credential`...

I propose the following:
- Add accessToken and apiBaseUrl to the source object. apiBaseUrl defaults to the US region.
- Add a setCartoAccessToken and setCartoApiBaseUrl. I add `Carto` because setAccessToken will be super generic when it's used in large applications. I know we could solve as alias, but I think it's better to force the default alias to carto

Reply:
Why just special case those two? In general what we are trying to do is to provide global `Source` configuration. One could argue there is value in allowing global configuration of `headers`, `clientId` etc.

In general I don't love the `setDefaultCredentials` pattern as it means all of the parameters (`accessToken`, `mapsUrl`) need to be typed as optional and we have no way to know at compile time if the user has forgotten to include something we require (for example access token). Also we rely on the user setting some global state which isn't great.

How about this as a pattern?

```javascript
import {SOURCE_DEFAULTS, CartoQuerySource, CartoVectorLayer} from '@deck.gl/carto';

const data = CartoQuerySource({
  ...SOURCE_DEFAULTS,
  accessToken: 'XXX',
  connectionName: 'bigquery',
  sqlQuery: 'select * from carto.tables.example'
});

new CartoVectorLayer({data});
```

In a larger codebase with different sources the user would define their own source:

```javascript
// AuthenticatedQuerySource.js
export default function AuthenticatedQuerySource(options) {
  return CartoQuerySource({...SOURCE_DEFAULTS, accessToken: 'XXX', connectionName: 'bigquery', ...options});
};

// Elsewhere
import AuthenticatedQuerySource from 'AuthenticatedQuerySource';

new CartoVectorLayer({
  data: AuthenticatedQuerySource({sqlQuery: 'select * from carto.tables.example'});
});
```

Reply:
Yes, what I wanted to propose is something more general rather than setDefaultCredentials. 

I like your approach